marlowe-runtime/tx/Language/Marlowe/Runtime/Transaction/Safety.hs

```diff
@@ -0,0 +1,389 @@
+  , EraHistory(..)
+  , ProtocolParameters(ProtocolParameters, protocolParamUTxOCostPerByte)
+  , SystemStart(..)
+  )
+import qualified Data.Map.Strict as M (Map, elems, empty, fromList, keys, map, mapKeys, singleton, size)
+import qualified Data.Set as S (singleton)
+import qualified Language.Marlowe.Core.V1.Merkle as V1 (MerkleizedContract(..))
+import qualified Language.Marlowe.Core.V1.Plate as V1 (extractAllWithContinuations)
+import qualified Language.Marlowe.Core.V1.Semantics as V1
+  (MarloweData(..), MarloweParams(..), TransactionInput(..), TransactionOutput(..))
+import qualified Language.Marlowe.Core.V1.Semantics.Types as V1 (Party(Address), State(..), Token(..))
+import qualified Language.Marlowe.Core.V1.Semantics.Types.Address as V1 (deserialiseAddress)
+import qualified Language.Marlowe.Runtime.ChainSync.Api as Chain
+  ( Address(..)
+  , AssetId(..)
+  , Assets(..)
+  , DatumHash(..)
+  , PolicyId(..)
+  , SlotNo(..)
+  , TokenName(..)
+  , Tokens(..)
+  , TransactionMetadata(..)
+  , TransactionOutput(..)
+  , TxOutRef
+  , UTxOs(..)
+  )
+import qualified Ouroboros.Consensus.BlockchainTime as Ouroboros (RelativeTime(..), mkSlotLength, toRelativeTime)
+import qualified Ouroboros.Consensus.HardFork.History as Ouroboros
+  ( Bound(..)
+  , EraEnd(..)
+  , EraParams(..)
+  , EraSummary(..)
+  , SafeZone(..)
+  , interpretQuery
+  , mkInterpreter
+  , summaryWithExactly
+  , wallclockToSlot
+  )
+import qualified Ouroboros.Consensus.Util.Counting as Ouroboros (Exactly(..))
+import qualified Ouroboros.Network.Block as Ouroboros (SlotNo(..))
+import qualified Plutus.V2.Ledger.Api as Plutus
+  (CurrencySymbol(..), DatumHash(..), POSIXTime(..), TokenName(..), fromBuiltin, toBuiltin)
+import qualified PlutusTx.AssocMap as AM (toList)
+
+
+-- FIXME: Relocate this definition when full support for Merkleization is added to Runtime.
+type Continuations v = M.Map Chain.DatumHash (Contract v)
+
+
+-- | No continuations for a contract.
+noContinuations
+  :: MarloweVersion v
+  -> Continuations v
+noContinuations MarloweV1 = M.empty
+
+
+-- | Map Plutus continuations into chain-sync continuations.
+remapContinuations
+  :: M.Map Chain.DatumHash contract
+  -> M.Map Plutus.DatumHash contract
+remapContinuations = M.mapKeys $ Plutus.DatumHash . Plutus.toBuiltin . Chain.unDatumHash
+
+
+-- | Compute a worst-case bound on the minimum UTxO value for a contract.
+minAdaUpperBound
+ :: Shelley.ProtocolParameters
+ -> MarloweVersion v
+ -> Contract v
+ -> Continuations v
+ -> Maybe Cardano.Lovelace
+minAdaUpperBound Shelley.ProtocolParameters{protocolParamUTxOCostPerByte} MarloweV1 contract continuations =
+  fromInteger
+    . (\utxoCostPerByte -> worstMinimumUtxo (toInteger utxoCostPerByte) contract $ remapContinuations continuations)
+    <$> protocolParamUTxOCostPerByte
+
+
+-- | Check a contract for design errors and ledger violations.
+checkContract
+  :: RoleTokensConfig
+  -> MarloweVersion v
+  -> Contract v
+  -> Continuations v
```

Comment:
Where does this error out if a continuation isn't found?

Reply:
This doesn't walk the continuations, so it won't error out if one is missing. I suppose we could add a check that all continuation hashes present in the contract are also keys in the map.

Reply:
I think that is crucial - or else we aren't doing a full analysis of the contract. The continuations could contain problems and the user would never know.

Reply:
Fixed in 7139176d82a3167d7c7ecbac9cf421457da675c7.

Reply:
Just to clarify, prior to 7139176d82a3167d7c7ecbac9cf421457da675c7 the checks were done on both the contract and its continuations, but there was not check to see that the map of continuations was complete.